src/DocumentFormat.OpenXml/Wordprocessing/Document.cs

```diff
@@ -3,20 +3,122 @@
             {
                 return true;
             }
 
             return base.StrictTranslateAttribute(qname, value);
         }
+
+        /// <summary>
+        /// Registers an existing w14:paraId value to ensure the uniqueness of values
+        /// generated later.
+        /// </summary>
+        /// <param name="value">The w14:paraId value.</param>
+        /// <returns>true, if <paramref name="value" /> did not exist; false, otherwise.</returns>
+        internal bool RegisterParagraphId(string value)
+        {
+            bool isAdded = ParagraphIds.Add(value.ToUpperInvariant());
+
+            if (!isAdded)
+            {
+                DuplicateParagraphIds.Add(value);
+            }
+
+            return isAdded;
+        }
+
+        /// <summary>
+        /// Registers an existing w14:paraId value to ensure the uniqueness of values
+        /// generated later.
+        /// </summary>
+        /// <param name="value">The w14:paraId value.</param>
+        /// <returns>true, if <paramref name="value" /> did not exist; false, otherwise.</returns>
+        /// <exception cref="ArgumentException">If the HexBinaryValue`s InnerText property value is null.</exception>
+        internal bool RegisterParagraphId(HexBinaryValue value)
+        {
+            if (value.InnerText is null)
+            {
+                throw new ArgumentException(@"Invalid HexBinaryValue: InnerText is null.", nameof(value));
+            }
+
+            return RegisterParagraphId(value.InnerText);
+        }
+
+        /// <summary>
+        /// Creates a secure random w14:paraId value but neither checks whether
+        /// the value is unique nor registers the value.
+        /// </summary>
+        /// <returns>A random w14:paraId value.</returns>
+        internal static HexBinaryValue CreateRandomParagraphId()
+        {
+            // We create a four-byte secure random number, noting that the first byte
+            // will become the most significant byte. Since we create non-zero bytes,
+            // the value will be greater than zero. We further ensure the value is less
+            // than 0x80000000 by masking the most significant byte.
+            var bytes = new byte[4];
+            Generator.GetNonZeroBytes(bytes);
+            bytes[0] &= 0x7f;
+
+            return HexBinaryValue.Create(bytes);
+        }
+
+        /// <summary>
+        /// Creates a secure random w14:paraId value that is unique within this Document.
+        /// </summary>
+        /// <remarks>
+        /// The previously existing w14:paraId values are registered when reading the DOM tree
+        /// from the OpenXmlPart, using the RegisterParagraphId() methods.
+        /// </remarks>
+        /// <returns>A random and unique w14:paraId value.</returns>
+        internal HexBinaryValue CreateUniqueParagraphId()
+        {
+            HexBinaryValue paragraphId = CreateRandomParagraphId();
+            while (!ParagraphIds.Add(paragraphId.InnerText!))
```

Comment:
Check if not null or make the ParagraphIds take `string?`

Reply:
What exactly do you mean?

The `HexBinaryValue` instances created by `CreateRandomParagraphId()` cannot be `null`. Visual Studio just thinks `InnerText` can be `null` because it is nullable. I'm not sure whether that makes sense (because a non-null `HexBinaryValue` should contain a valid, non-null value and, therefore, non-null `InnerText`).

The method creates a `HexBinaryValue` because the downstream code expects a `HexBinaryValue`.

Reply:
Yes, it *should* contain a non-null value at this point, but nothing is guaranteeing that since `HexBinaryValue` *may* contain a null `InnerText`. Can you verify it's null and throw if not, or allow the hashset to take a null? If you just need the innertext, have the `CreateRandomParagraphId` create that instead of a HexBinaryValue

Reply:
**<w:rant>**

It _will definitely_ contain a nun-noll value at this point. The `HexBinaryValue` returned as follows is guaranteed to contain a non-null `TextValue` (assigned by the constructor) and, therefore, non-null `InnerText` (and `Value`):

```csharp
return HexStringFactory.Create(bytes);
```

The corresponding unit test in `HexBinaryValueTests` establishes that fact (for empty byte arrays and those with one or more elements), although it is not easy to see because the unit test methods use either the `Value` or the `InnerText` property, both of which delegate to `TextValue`, though. So, to be honest, I was confused as to which property I'm supposed to use (i.e., `Value` or `InnerText`; the documentation tells you to not use `TextValue`).

A `HexBinaryValue` instance _may only_ contain a null `InnerText` (or `TextValue` or `Value`) if I am passed a `HexBinaryValue` that was created outside of my control. In this case, however, I am fully in control. The `HexBinaryValue` value is created by a utility method of the same class. Having to check for `null` in this case would be both weird and either:

- consume CPU cycles for no good reason or
- hopefully, be optimized away by the compiler.

Can you explain to me why the `HexBinaryValue` constructor even takes a `null` value? Given an empty `byte` array or `ReadOnlySpan<byte>`, the two `HexBinaryValue.Create()` methods will `return new HexBinaryValue(string.Empty)`. This is either inconsistent or the way to go unless you go one step further and demand a `HexBinaryValue` be an actual hex binary value, which would be `00` at least (i.e., same rules as for `int`, `uint`, etc.).

**</w:rant>**

**<w:wayForward>**

Having said that, I have taken a different route and indeed created a `ParagraphIdService` (as you suggested) that does not make use of the `HexBinaryValue` class anymore (to be reusable also for the Linq to XML use case). Since that service still requires code that produces a hex string from a byte array, I've carved out the code used by the `HexBinaryValue` class and created a `HexStringFactory` class with `Create()` methods that create `string` rather than `HexBinaryValue` instances. Both the `HexBinaryValue.Create(ReadOnlySpan<byte> bytes)` method and the `ParagraphIdService` class use the `HexStringFactory` to create hex strings. The (`public` or `internal`) `HexBinaryValue.Create(ReadOnlySpan<byte> bytes)` then simply looks like this:

```csharp
static HexBinaryValue Create(ReadOnlySpan<byte> bytes)
{
    return new(HexStringFactory.Create(bytes));
}
```

I will soon push new commits with those changes.

**</w:wayForward>**

Reply:
My point is nothing about behavior, but that I'd rather not go around the null reference checks. Maybe the annotations can be updated to something that will tell the compiler that it would be non-null at that point to *fix* this.

As for why it takes a null - it was like that from before I got involved. Potentially a reasonable breaking change at somepoint, though. I'll create an issue to track that.

Reply:
See #1009

Reply:
Thanks, @twsouthwick. You asked exactly the right question in #1009. If it's a breaking change, let's go for it in the next major release.